### PR TITLE
GH-41818: [C++][Parquet] normalize dictionary encoding to use RLE_DICTIONARY

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

There're some points:

1. https://github.com/apache/arrow/blob/main/cpp/src/parquet/encoding.cc#L444-L445 . encoding is not passed in Encoder
2. But, it's RLE in decoder: https://github.com/apache/arrow/blob/main/cpp/src/parquet/encoding.cc#L1607 it will be detect and normalized in other place, like:
3. https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L876

We'd better unifying them

### What changes are included in this PR?

Unify dict encoding to `RLE_DICTIONARY`.

### Are these changes tested?

No need

### Are there any user-facing changes?

No

* GitHub Issue: #41818

### CODE REVIEW

**Comment by mapleFU** [comment]
cc @wgtmac @pitrou @ClifHouck

File: cpp/src/parquet/encoding.cc [hunk 1]
```diff
@@ -442,12 +442,12 @@ class DictEncoderImpl : public EncoderImpl, virtual public DictEncoder<DType> {
   constexpr static int32_t kDataPageBitWidthBytes = 1;
 
   explicit DictEncoderImpl(const ColumnDescriptor* desc, MemoryPool* pool)
-      : EncoderImpl(desc, Encoding::PLAIN_DICTIONARY, pool),
+      : EncoderImpl(desc, Encoding::RLE_DICTIONARY, pool),
```
Comment (wgtmac) [inline cpp/src/parquet/encoding.cc:445]: I'm not sure if this is a minor breaking change to those legacy readers which stick to `PLAIN_DICTIONARY`. cc @jorisvandenbossche

Comment (mapleFU) [reply cpp/src/parquet/encoding.cc:445]: IMO this is only a Parquet internal API, maybe user shouldn't depend on this?

File: cpp/src/parquet/properties.h [hunk 2]
```diff
@@ -1180,4 +1180,10 @@ struct ArrowWriteContext {
 PARQUET_EXPORT
 std::shared_ptr<ArrowWriterProperties> default_arrow_writer_properties();
 
+// PLAIN_DICTIONARY is deprecated but used to be used as a dictionary index
```
Comment (pitrou) [inline cpp/src/parquet/properties.h:1183]: Why put this in `properties.h`? Perhaps `types.h` would be more appropriate.

Comment (mapleFU) [reply cpp/src/parquet/properties.h:1183]: Aha, previously `dictionary_index_encoding()` is in `properties.h`. I'll move it

Comment (mapleFU) [reply cpp/src/parquet/properties.h]: Moved to `types.h` now

File: cpp/src/parquet/encoding.cc (hunk 1 above)
Comment (mapleFU) [reply cpp/src/parquet/encoding.cc:445]: It doesn't change what we written to storage? The storage would final decided by `Properties::dictionary_index_encoding()`

Comment (wgtmac) [reply cpp/src/parquet/encoding.cc:445]: Yes, you're right. I forgot this.

File: cpp/src/parquet/types.h [hunk 3]
```diff
@@ -811,4 +811,10 @@ PARQUET_EXPORT SortOrder::type GetSortOrder(ConvertedType::type converted,
 PARQUET_EXPORT SortOrder::type GetSortOrder(
     const std::shared_ptr<const LogicalType>& logical_type, Type::type primitive);
 
+/// PLAIN_DICTIONARY is deprecated but used to be used as a dictionary index
+/// encoding.
```
Comment (pitrou) [inline cpp/src/parquet/types.h:815]: Please make sure the docstring explains what the function does.
If you want to explain implementation details, regular comments (`//`) should be used.

Comment (mapleFU) [reply cpp/src/parquet/types.h]: Ah I didn't know this, changed now

**Comment by pitrou** [comment]
Can you rebase to get up-to-date CI results?

**Review by pitrou — APPROVED** [review]

**Comment by mapleFU** [comment]
done

**Comment by pitrou** [comment]
Thank you @mapleFU !

**Pull request merged**